odoo/addons/base/models/ir_http.py

```diff
@@ -464,12 +467,19 @@ def binary_content(self, xmlid=None, model='ir.attachment', id=None, field='raw'
         :param str default_mimetype: default mintype if no mintype found
         :param str access_token: optional token for unauthenticated access
                                  only available  for ir.attachment
+        :param str allow_xaccel: optional argument to bypass xaccel NGINX serving
+                                and fetch raw data instead of only a header
         :returns: (status, headers, content)
         """
         record, status = self._get_record_and_check(xmlid=xmlid, model=model, id=id, field=field, access_token=access_token)
+        xaccel_header = False
 
         if not record:
-            return (status or 404, [], None)
+            return status or 404, [], None
+
+        if odoo.tools.config['xaccel'] and hasattr(record, "store_fname") and record.store_fname and allow_xaccel:
+            # Uses NGINX location /xaccel/ to serve files through NGINX instead of reading them
+            xaccel_header = ('X-Accel-Redirect', f"/xaccel/{record.store_fname}")
```

Comment:
- Passing this flag through 15 layers of calls to end up here seems weird. Even more so as we apparently... straight up dont send content in that case?
- Hardcoding the access path seems odd, even more so when it's neither the path nginx uses in their documentation nor one that is namespaced for odoo.

Reply:
> * Passing this flag through 15 layers of calls to end up here seems weird. Even more so as we apparently... straight up dont send content in that case?

Yes, the X-Accel-Redirect header will point to the filestore and the content will be served by NGINX. If enabled, we want this for every possible file, since it's always better to serve it with NGINX than serving it through Odoo workers. I agree that it's weird that it has to be passed through that many layers to end up here. It was already discussed with @fpodoo and @odony but ended up being left as is. Argument for was that it followed current structure and there would be a task in the near future to change that structure. Arguments against were that it was illogical and the check should be somewhere lower as you insinuate.

> * Hardcoding the access path seems odd, even more so when it's neither the path nginx uses in their documentation nor one that is namespaced for odoo.

I haven't seen the Odoo SaaS configuration, so I had no idea there was a specific namespace for NGINX locations linked to Odoo. It seemed logical to me that a location called xaccel would be available and could be used for the redirects without causing confusion. What would you suggest as a name ?

Reply:
> * Passing this flag through 15 layers of calls to end up here seems weird. 

Yes I had an issue with that spaghetti problem, and the dilution of responsibility between the different places of the stack:

→ `_content_image()` decides whether X-Accel is possible and tells `binary_content()`
  ↳ `binary_content()` decides whether the header should be generated, then passes it down:
    ↳ `_binary_record_content()` receives it to optimize out(?) the binary data fetch
    ↳ `_binary_set_headers()` sets the already prepared header

Ideally the decision and action should happen all in the same spot. But that's not easy with current structure. And refactoring this puts us back to the drawing board about what piece of code is responsible for what.

Here's a thought: @kzhekov, what if we used a trick to obtain a better trade-off?
- `_binary_record_content()` could return a lazy content object, that includes the data source in a closure. It would only fetch the bytes for real when you consume them. This way if it's discarded downstream it won't ever be fetched.
- `_content_image()` could then be in charge of all the decisions and actions for X-Accel. And yes, `_binary_set_headers` won't be the only one setting headers, but that's not too bad.

This way we don't need to modify the structure, we keep the optimizations, and we don't need to pass extra parameters anywhere.

Reply:
This sounds like a good balance between the existing structure and possible optimizations. I'll try to adapt it to the structure you describe the best I can 😄 Thanks !

Reply:
> I haven't seen the Odoo SaaS configuration, so I had no idea there was a specific namespace for NGINX locations linked to Odoo.

There isn't. But the server here needs to use a prefix that makes sense and can be easily mapped to a relevant internal nginx location. Especially if we don't want to make that prefix configurable in the server config.

Perhaps you could use `/protected_odoo_filestore/` as the prefix? It builds on the `/protected_files` from the nginx examples, and hints that you're supposed to point it to a local Odoo filestore. Then we'll use that in the doc section where we give the nginx config example.
BTW the filestore path is db-specific, so the nginx internal location will need to point *inside* the filestore, to the db-specific root. That's *good* and *important* because it means we won't be accidentally sharing files cross-database for multi-tenant setups.

Reply:
> BTW the filestore path is db-specific, so the nginx internal location will need to point _inside_ the filestore, to the db-specific root. That's _good_ and _important_ because it means we won't be accidentally sharing files cross-database for multi-tenant setups.

Should I add that db specificity to the X-Accel-Redirect header, so that it uses something like `/my_odoo_db/protected_odoo_filestore/path_to_file` instead of the current `/protected_odoo_filestore/path_to_file` or is it something that can be taken care of entirely on the NGINX side ? Locally, I just hardcoded the path to my own database but that's of course not what is done on the actual servers 😅

Reply:
> I haven't seen the Odoo SaaS configuration, so I had no idea there was a specific namespace for NGINX locations linked to Odoo.

There is not one, but if we're hard-coding a made-up location surely it would make sense for that location to mention odoo in some way, in order to make its purpose self-documenting?

> It seemed logical to me that a location called xaccel would be available and could be used for the redirects without causing confusion. What would you suggest as a name ?

Maybe passing one explicitely? Otherwise either the ones nginx documentation mentions (e.g. `/protected` or `/protected_files`) or a variation thereof mentioning odoo (e.g. `/protected-odoo` or `/odoo-filestore`).

Reply:
Yes, I went with protected_odoo_filestore for now, it's definitely clearer than just xaccel 😅 Thanks !

Reply:
> Should I add that db specificity to the X-Accel-Redirect header, so that it uses something like `/my_odoo_db/protected_odoo_filestore/path_to_file` instead of the current `/protected_odoo_filestore/path_to_file`

No, that was my point. Otherwise database `foo` would be able to blindly forge a X-Accel-Redirect header that points to database `bar`'s filestore, causing a security problem.

To avoid that, nginx's config must not trust that header, it has to verify that the database prefix matches what is expected, based on the `Host` which served the request. (e.g. "Host was foo.com -> database should be `foo`"). Of course at that point, nginx can simply use the db prefix that is has determined.
So it's better not to return the db name at all, to make sure absentminded sysadmins won't trust it by mistake.

I suggest putting a comment in the code to explain that it's really done on purpose.

---

On a slightly related note, the example nginx config should avoid path traversal vulnerabilities by reminding admins that the `location` directives must end with a slash `/`, especially if they include an `alias` directive to implement the internal X-Accel.

Reply:
There is a WIP of the raw lazy loading wrapper, still have to test it and see if it does what it's supposed to in all expected situations 😰